File: src/main/java/com/sgx/icms/service/SurveyorService.java
package com.sgx.icms.service;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.Arrays;
import java.util.List;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.sgx.icms.dao.AssessmentDao;
import com.sgx.icms.dao.ClaimDao;
import com.sgx.icms.db.Db;
import com.sgx.icms.domain.Assessment;
import com.sgx.icms.domain.AssessmentComponent;
import com.sgx.icms.domain.AuditLog;
import com.sgx.icms.domain.Claim;
import com.sgx.icms.domain.ClaimStatus;
import com.sgx.icms.web.support.Paged;
import com.sgx.icms.web.support.SessionUser;

/**
 * Surveyor use-cases: list claims assigned to the surveyor (ownership-enforced),
 * load a claim's assessment, and submit a damage assessment with a component
 * breakdown. The monetary maths (gross → depreciation → net payable) is computed
 * server-side — never trusted from the client.
 */
public class SurveyorService {

    private static final Logger LOG = LoggerFactory.getLogger(SurveyorService.class);

    private static final List<String> PENDING_STATUSES = Arrays.asList(ClaimStatus.SURVEY_SCHEDULED);
    private static final List<String> ASSESSED_STATUSES = Arrays.asList(
            ClaimStatus.UNDER_ASSESSMENT, ClaimStatus.PENDING_APPROVAL, ClaimStatus.APPROVED,
            ClaimStatus.SETTLEMENT_PROCESSING, ClaimStatus.SETTLED, ClaimStatus.CLOSED);

    private final ClaimDao claimDao = new ClaimDao();
    private final AssessmentDao assessmentDao = new AssessmentDao();
    private final NotificationService notifications = new NotificationService();
    private final AuditService audit = new AuditService();

    public Paged<Claim> assignedClaims(long surveyorId, int page, int size) {
        int offset = (page - 1) * size;
        return Db.withConnection(conn -> {
            long total = claimDao.countBySurveyor(conn, surveyorId);
            List<Claim> items = claimDao.findBySurveyor(conn, surveyorId, size, offset);
            return new Paged<>(items, page, size, total);
        });
    }

    /** [total, pendingSurvey, assessed] for the surveyor dashboard. */
    public long[] counts(long surveyorId) {
        return Db.withConnection(conn -> new long[] {
                claimDao.countBySurveyor(conn, surveyorId),
                claimDao.countBySurveyorInStatuses(conn, surveyorId, PENDING_STATUSES),
                claimDao.countBySurveyorInStatuses(conn, surveyorId, ASSESSED_STATUSES)
        });
    }

    /** Claim only if assigned to this surveyor (authorisation). */
    public Claim getAssignedClaim(long surveyorId, long claimId) {
        Claim c = Db.withConnection(conn -> claimDao.findById(conn, claimId));
        if (c == null || c.getSurveyorId() == null || c.getSurveyorId() != surveyorId) {
            return null;
        }
        return c;
    }

    public Assessment latestAssessment(long claimId) {
        return Db.withConnection(conn -> assessmentDao.findByClaim(conn, claimId));
    }

    public List<AssessmentComponent> components(long assessmentId) {
        return Db.withConnection(conn -> assessmentDao.findComponents(conn, assessmentId));
    }

    /**
     * Persists a submitted assessment. Computes gross (from component repair costs),
     * depreciation amount, and net payable server-side. Moves the claim to
     * UNDER_ASSESSMENT and notifies the handling agent.
     *
     * @throws IllegalArgumentException on invalid input
     * @throws IllegalStateException if the claim isn't assigned to this surveyor / wrong state
     */
    public void submitAssessment(SessionUser surveyor, long claimId, Assessment input,
                                 List<AssessmentComponent> components, String ip) {
        BigDecimal gross = sumRepairCosts(components);
        if (gross.signum() <= 0 && nz(input.getGrossAssessed()).signum() <= 0) {
            throw new IllegalArgumentException("Add at least one component (or a gross assessed amount).");
        }
        if (gross.signum() <= 0) {
            gross = nz(input.getGrossAssessed());
        }

        BigDecimal deductible = nz(input.getPolicyDeductible());
        BigDecimal deprPct = nz(input.getDepreciationPct());
        BigDecimal salvage = nz(input.getSalvageValue());
        BigDecimal deprAmt = gross.multiply(deprPct).divide(BigDecimal.valueOf(100), 2, RoundingMode.HALF_UP);
        BigDecimal net = gross.subtract(deductible).subtract(deprAmt).subtract(salvage);
        if (net.signum() < 0) {
            net = BigDecimal.ZERO;
        }

        input.setGrossAssessed(gross);
        input.setDepreciationAmt(deprAmt);
        input.setNetPayable(net);
        input.setSurveyorId(surveyor.getId());
        input.setClaimId(claimId);
        input.setStatus(Assessment.STATUS_SUBMITTED);

        final BigDecimal netFinal = net;
        Db.inTransaction(conn -> {
            Claim claim = claimDao.findById(conn, claimId);
            if (claim == null || claim.getSurveyorId() == null || claim.getSurveyorId() != surveyor.getId()) {
                throw new IllegalStateException("This claim is not assigned to you.");
            }
            if (ClaimStatus.isTerminal(claim.getStatus())) {
                throw new IllegalStateException("This claim is closed.");
            }
            Assessment existing = assessmentDao.findByClaim(conn, claimId);
            if (existing != null && Assessment.STATUS_SUBMITTED.equals(existing.getStatus())) {
                throw new IllegalStateException("An assessment has already been submitted for this claim.");
            }

            long assessmentId = assessmentDao.insert(conn, input);
            if (components != null) {
                for (AssessmentComponent c : components) {
                    c.setAssessmentId(assessmentId);
                    assessmentDao.insertComponent(conn, c);
                }
            }
            claimDao.updateStatus(conn, claimId, ClaimStatus.UNDER_ASSESSMENT);

            if (claim.getAgentId() != null) {
                notifications.notifyUser(conn, claim.getAgentId(), "ACTION",
                        "Survey report uploaded for claim " + claim.getClaimNo()
                        + " (net payable ₹ " + netFinal + ").");
            }
            writeAudit(conn, surveyor, "ASSESSMENT_SUBMITTED", claim.getClaimNo(), ip);
            LOG.info("Surveyor {} submitted assessment for claim {} (net {})",
                    surveyor.getUsername(), claim.getClaimNo(), netFinal);
            return null;
        });
    }

    private static BigDecimal sumRepairCosts(List<AssessmentComponent> components) {
        BigDecimal sum = BigDecimal.ZERO;
        if (components != null) {
            for (AssessmentComponent c : components) {
                sum = sum.add(nz(c.getRepairCost()));
            }
        }
        return sum;
    }

    private static BigDecimal nz(BigDecimal v) {
        return v == null ? BigDecimal.ZERO : v;
    }

    private void writeAudit(java.sql.Connection conn, SessionUser actor, String action, String entity, String ip) {
        AuditLog a = new AuditLog();
        a.setUserId(actor.getId());
        a.setUsername(actor.getUsername());
        a.setRole(actor.getRole());
        a.setAction(action);
        a.setEntity(entity);
        a.setIpAddress(ip);
        a.setResult(AuditLog.RESULT_SUCCESS);
        audit.record(conn, a);
    }
}

File: backend/src/services/surveyorService.js
'use strict';

const { knex } = require('../db/knex');
const { withTransaction } = require('../db/tx');
const claimRepo = require('../repositories/claimRepo');
const assessmentRepo = require('../repositories/assessmentRepo');
const documentService = require('./documentService');
const notifications = require('./notificationService');
const audit = require('./auditService');
const claimStatus = require('../domain/claimStatus');
const { withStatusView } = require('./claimService');
const { Decimal, dec, toAmount } = require('../utils/money');
const { paged } = require('../utils/paging');
const { NotFoundError, ValidationError, ConflictError } = require('../utils/errors');
const logger = require('../config/logger');

const ASSESSMENT_SUBMITTED = 'SUBMITTED';

const PENDING_STATUSES = [claimStatus.SURVEY_SCHEDULED];
const ASSESSED_STATUSES = [
  claimStatus.UNDER_ASSESSMENT,
  claimStatus.PENDING_APPROVAL,
  claimStatus.APPROVED,
  claimStatus.SETTLEMENT_PROCESSING,
  claimStatus.SETTLED,
  claimStatus.CLOSED,
];

async function assignedClaims(surveyorId, page, size, offset) {
  const [total, rows] = await Promise.all([
    claimRepo.countBySurveyor(knex, surveyorId),
    claimRepo.findBySurveyor(knex, surveyorId, size, offset),
  ]);
  return paged(rows.map(withStatusView), page, size, total);
}

/** {total, pendingSurvey, assessed} for the surveyor dashboard. */
async function counts(surveyorId) {
  const [total, pendingSurvey, assessed] = await Promise.all([
    claimRepo.countBySurveyor(knex, surveyorId),
    claimRepo.countBySurveyorInStatuses(knex, surveyorId, PENDING_STATUSES),
    claimRepo.countBySurveyorInStatuses(knex, surveyorId, ASSESSED_STATUSES),
  ]);
  return { total, pendingSurvey, assessed };
}

/** Claim only if assigned to this surveyor (authorisation). */
async function getAssignedClaim(surveyorId, claimId) {
  const c = await claimRepo.findById(knex, claimId);
  if (!c || c.surveyorId == null || Number(c.surveyorId) !== Number(surveyorId)) return null;
  return c;
}

/** Assessment screen model: claim + latest assessment + components + documents. */
async function assessScreen(surveyorId, claimId) {
  const claim = await getAssignedClaim(surveyorId, claimId);
  if (!claim) return null;
  const assessment = await assessmentRepo.findByClaim(knex, claimId);
  const components = assessment ? await assessmentRepo.findComponents(knex, assessment.id) : [];
  const documents = await documentService.forClaim(claimId);
  return { claim: withStatusView(claim), assessment, components, documents };
}

/**
 * Persists a submitted assessment. Computes gross/depreciation/net payable
 * server-side (never trusts the client). Moves the claim to UNDER_ASSESSMENT and
 * notifies the handling agent. Port of SurveyorService.submitAssessment.
 */
async function submitAssessment(surveyor, claimId, input, components, ip) {
  let gross = sumRepairCosts(components);
  if (gross.lessThanOrEqualTo(0) && dec(input.grossAssessed).lessThanOrEqualTo(0)) {
    throw new ValidationError('Add at least one component (or a gross assessed amount).');
  }
  if (gross.lessThanOrEqualTo(0)) gross = dec(input.grossAssessed);

  const deductible = dec(input.policyDeductible);
  const deprPct = dec(input.depreciationPct);
  const salvage = dec(input.salvageValue);
  const deprAmt = gross.times(deprPct).dividedBy(100).toDecimalPlaces(2, Decimal.ROUND_HALF_UP);
  let net = gross.minus(deductible).minus(deprAmt).minus(salvage);
  if (net.isNegative()) net = new Decimal(0);

  const record = {
    ...input,
    grossAssessed: gross.toFixed(2),
    depreciationAmt: deprAmt.toFixed(2),
    policyDeductible: toAmount(deductible),
    depreciationPct: toAmount(deprPct),
    salvageValue: toAmount(salvage),
    netPayable: net.toFixed(2),
    surveyorId: Number(surveyor.id),
    claimId,
    status: ASSESSMENT_SUBMITTED,
  };
  const netFinal = net.toFixed(2);

  await withTransaction(async (trx) => {
    const claim = await claimRepo.findById(trx, claimId);
    if (!claim || claim.surveyorId == null || Number(claim.surveyorId) !== Number(surveyor.id)) {
      throw new NotFoundError('This claim is not assigned to you.');
    }
    if (claimStatus.isTerminal(claim.status)) throw new ConflictError('This claim is closed.');
    const existing = await assessmentRepo.findByClaim(trx, claimId);
    if (existing && existing.status === ASSESSMENT_SUBMITTED) {
      throw new ConflictError('An assessment has already been submitted for this claim.');
    }

    const assessmentId = await assessmentRepo.insert(trx, record);
    for (const c of components || []) {
      await assessmentRepo.insertComponent(trx, { ...c, assessmentId });
    }
    await claimRepo.updateStatus(trx, claimId, claimStatus.UNDER_ASSESSMENT);

    if (claim.agentId != null) {
      await notifications.notifyUser(
        trx,
        Number(claim.agentId),
        'ACTION',
        `Survey report uploaded for claim ${claim.claimNo} (net payable ₹ ${netFinal}).`
      );
    }
    await audit.record(trx, {
      userId: Number(surveyor.id),
      username: surveyor.username,
      role: surveyor.role,
      action: 'ASSESSMENT_SUBMITTED',
      entity: claim.claimNo,
      ipAddress: ip,
      result: 'SUCCESS',
    });
    logger.info({ surveyor: surveyor.username, claimNo: claim.claimNo, net: netFinal }, 'Assessment submitted');
  });

  return { netPayable: netFinal };
}

function sumRepairCosts(components) {
  let sum = new Decimal(0);
  for (const c of components || []) sum = sum.plus(dec(c.repairCost));
  return sum;
}

module.exports = {
  assignedClaims,
  counts,
  getAssignedClaim,
  assessScreen,
  submitAssessment,
};
